runtime/include/mem/jemalloc/chpl-mem-impl.h

```diff
@@ -23,41 +23,60 @@
 
 // jemalloc.h references the token "malloc" (but not the actual function) and
 // our warning macros mess up jemalloc's use of it.
+// Also the CHPL_JE_ macro needs malloc etc to not be #defined
 #include "chpl-mem-no-warning-macros.h"
-#include "jemalloc.h"
-#include "chpl-mem-warning-macros.h"
+
+// Set up a #define to control jemalloc.h
+#ifndef CHPL_JEMALLOC_PREFIX
+// Use the je_ prefix
+#define CHPL_JEMALLOC_PREFIX je_
+// Ask jemalloc.h to provide the je_ functions
+// (i.e. je_malloc rather than just malloc)
+#define JEMALLOC_NO_DEMANGLE
+#endif
+
+#include "jemalloc/jemalloc.h"
+
 #include "chpltypes.h"
 
 #define MALLOCX_NO_FLAGS 0
 
+// Default configuration of jemalloc names the routines
+// like je_malloc, je_free, etc; but the Chapel configuration
+// uses the prefix chpl_je_. Use the CHPL_JE_ macro to
+// paste the appropriate prefix (je_ or chpl_je_).
+#define CHPL_JE_TOK_PASTE_INTERNAL(a, b) a ## b
+#define CHPL_JE_TOK_PASTE(a, b) CHPL_JE_TOK_PASTE_INTERNAL(a,b)
+#define CHPL_JE_(NAME) CHPL_JE_TOK_PASTE(CHPL_JEMALLOC_PREFIX, NAME)
+
 
 static inline void* chpl_calloc(size_t n, size_t size) {
-  return chpl_je_calloc(n,size);
+  return CHPL_JE_(calloc)(n,size);
```

Comment:
Good heavens, that's ugly.  Can we at least do

    #define CHPL_JE_CALLOC CHPL_JE_(calloc)`

and similarly for all the other functions, and then use `CHPL_JE_CALLOC` everywhere we want the symbol?